crates/uv-publish/src/trusted_publishing.rs

```diff
@@ -0,0 +1,154 @@
+//! Trusted publishing (via OIDC) with GitHub actions.
+
+use reqwest::{header, StatusCode};
+use reqwest_middleware::ClientWithMiddleware;
+use serde::{Deserialize, Serialize};
+use std::env;
+use std::env::VarError;
+use thiserror::Error;
+use tracing::{debug, trace};
+use url::Url;
+
+#[derive(Debug, Error)]
+pub enum TrustedPublishingError {
+    #[error(transparent)]
+    Var(#[from] VarError),
+    #[error(transparent)]
+    Url(#[from] url::ParseError),
+    #[error(transparent)]
+    Reqwest(#[from] reqwest::Error),
+    #[error(transparent)]
+    ReqwestMiddleware(#[from] reqwest_middleware::Error),
+    #[error(transparent)]
+    SerdeJson(#[from] serde_json::error::Error),
+    #[error(
+        "PyPI returned error code {0}, is trusted publishing correctly configured?\nResponse: {1}"
+    )]
+    Pypi(StatusCode, String),
+}
+
+/// The response from querying `https://pypi.org/_/oidc/audience`.
+#[derive(Deserialize)]
+struct Audience {
+    audience: String,
+}
+
+/// The response from querying `$ACTIONS_ID_TOKEN_REQUEST_URL&audience=pypi`.
+#[derive(Deserialize)]
+struct OidcToken {
+    value: String,
+}
+
+/// The body for querying `$ACTIONS_ID_TOKEN_REQUEST_URL&audience=pypi`.
+#[derive(Serialize)]
+struct MintTokenRequest {
+    token: String,
+}
+
+/// The response from querying `$ACTIONS_ID_TOKEN_REQUEST_URL&audience=pypi`.
+#[derive(Deserialize)]
+struct PublishToken {
+    token: String,
+}
+
+/// Returns the short-lived token to use for uploading.
+pub(crate) async fn get_token(
+    registry: &Url,
+    client: &ClientWithMiddleware,
+) -> Result<String, TrustedPublishingError> {
+    // If this fails, we can skip the audience request.
+    let oidc_token_request_token = env::var("ACTIONS_ID_TOKEN_REQUEST_TOKEN")?;
+
+    // Request 1: Get the audience
+    let audience = get_audience(registry, client).await?;
+
+    // Request 2: Get the OIDC token from GitHub.
+    let oidc_token = get_oidc_token(&audience, &oidc_token_request_token, client).await?;
+
+    // Request 3: Get the publishing token from PyPI.
+    let publish_token = get_publish_token(registry, &oidc_token, client).await?;
+
+    debug!("Received token, using trusted publishing");
+
+    // Tell GitHub Actions to mask the token in any console logs.
+    #[allow(clippy::print_stdout)]
+    {
+        // If we successfully obtained a token, we know we must be in GitHub Actions, so it's safe
+        // to use GitHub Actions commands.
+        println!("::add-mask::{}", &publish_token);
+    }
+
+    Ok(publish_token)
+}
+
+async fn get_audience(
+    registry: &Url,
+    client: &ClientWithMiddleware,
+) -> Result<String, TrustedPublishingError> {
+    // `pypa/gh-action-pypi-publish` uses `netloc` (RFC 1808), which is deprecated for authority
+    // (RFC 3986).
+    let audience_url = Url::parse(&format!("https://{}/_/oidc/audience", registry.authority()))?;
```

Comment:
Nit: you could DRY up this URL builder which is also used below.

Reply:
I had looked for this, but i couldn't find anything in the url crate that would allow me to build an url from an authority.